### PR TITLE
Feature/#419 tsam 2 level approach

### PR DESCRIPTION
**Pull request by KathiEsterl**
In this branch, I implemented the **2-level-approach** used in my master thesis. 

Additionally, I changed the order of the methods **skip_snapshots** and **snapshot_clustering** as it makes more sense to use them in this way, if someone would like to combine those.

### CODE REVIEW

**Comment by ClaraBuettner** [comment]
I tried to run this branch (5 segments of 24h) but I run into the following error:

```
Traceback (most recent call last):

  File "/home/clara/etrago_pypsa19/eTraGo/etrago/appl.py", line 512, in <module>
    etrago = run_etrago(args, json_path=None)

  File "/home/clara/etrago_pypsa19/eTraGo/etrago/appl.py", line 482, in run_etrago
    etrago.snapshot_clustering()

  File "/home/clara/etrago_pypsa19/eTraGo/etrago/cluster/snapshot.py", line 60, in snapshot_clustering
    csv_export = self.args['csv_export'] # can be deleted later, just helpful to see how time steps were clustered

  File "/home/clara/etrago_pypsa19/eTraGo/etrago/cluster/snapshot.py", line 279, in run
    update_data_frames(network, cluster_weights, dates, hours, timeseries, segmentation)

  File "/home/clara/etrago_pypsa19/eTraGo/etrago/cluster/snapshot.py", line 329, in update_data_frames
    index = timeseries.index.get_level_values(0))

  File "/home/clara/etrago_pypsa19/venv_pypsa19/lib/python3.7/site-packages/pypsa/components.py", line 440, in snapshot_weightings
    assert df.index.equals(self.snapshots), "Weightings not defined for all snapshots."

AssertionError: Weightings not defined for all snapshots.
```

Have you ever seen this before?

**Comment by ClaraBuettner** [comment]
The clustering is running now, but I got this warning: 

```
Snapshot clustering to 2 days
Using extreme period method: None
weight of "residual_load" set to the minmal tolerable weighting
/home/clara/etrago_pypsa19/venv_pypsa19/lib/python3.7/site-packages/tsam/timeseriesaggregation.py:982: UserWarning: Something went wrong: At least one maximal value of the aggregated time series exceeds the maximal value the input time series
  warnings.warn("Something went wrong: At least one maximal value of the aggregated time series exceeds the maximal value the input time series")
/home/clara/etrago_pypsa19/venv_pypsa19/lib/python3.7/site-packages/tsam/timeseriesaggregation.py:984: UserWarning: Something went wrong: At least one minimal value of the aggregated time series exceeds the minimal value the input time series
  warnings.warn("Something went wrong: At least one minimal value of the aggregated time series exceeds the minimal value the input time series")
DatetimeIndex(['2011-01-01 00:00:00', '2011-01-01 01:00:00',
               '2011-01-01 02:00:00', '2011-01-01 03:00:00',
               '2011-01-01 04:00:00', '2011-01-01 05:00:00',
               '2011-01-01 06:00:00', '2011-01-01 07:00:00',
               '2011-01-01 08:00:00', '2011-01-01 09:00:00',
               '2011-01-01 10:00:00', '2011-01-01 11:00:00',
               '2011-01-01 12:00:00', '2011-01-01 13:00:00',
               '2011-01-01 14:00:00', '2011-01-01 15:00:00',
               '2011-01-01 16:00:00', '2011-01-01 17:00:00',
               '2011-01-01 18:00:00', '2011-01-01 19:00:00',
               '2011-01-01 20:00:00', '2011-01-01 21:00:00',
               '2011-01-01 22:00:00', '2011-01-01 23:00:00',
               '2011-01-03 00:00:00', '2011-01-03 01:00:00',
               '2011-01-03 02:00:00', '2011-01-03 03:00:00',
               '2011-01-03 04:00:00', '2011-01-03 05:00:00',
               '2011-01-03 06:00:00', '2011-01-03 07:00:00',
               '2011-01-03 08:00:00', '2011-01-03 09:00:00',
               '2011-01-03 10:00:00', '2011-01-03 11:00:00',
               '2011-01-03 12:00:00', '2011-01-03 13:00:00',
               '2011-01-03 14:00:00', '2011-01-03 15:00:00',
               '2011-01-03 16:00:00', '2011-01-03 17:00:00',
               '2011-01-03 18:00:00', '2011-01-03 19:00:00',
               '2011-01-03 20:00:00', '2011-01-03 21:00:00',
               '2011-01-03 22:00:00', '2011-01-03 23:00:00'],
              dtype='datetime64[ns]', name='snapshot', freq=None)
INFO:pypsa.opf:Performed preliminary steps
```
I tried to cluster the first 72h to 2 days.

**Comment by KathiEsterl** [comment]
> The clustering is running now, but I got this warning:
> 
> ```
> Snapshot clustering to 2 days
> Using extreme period method: None
> weight of "residual_load" set to the minmal tolerable weighting
> /home/clara/etrago_pypsa19/venv_pypsa19/lib/python3.7/site-packages/tsam/timeseriesaggregation.py:982: UserWarning: Something went wrong: At least one maximal value of the aggregated time series exceeds the maximal value the input time series
>   warnings.warn("Something went wrong: At least one maximal value of the aggregated time series exceeds the maximal value the input time series")
> /home/clara/etrago_pypsa19/venv_pypsa19/lib/python3.7/site-packages/tsam/timeseriesaggregation.py:984: UserWarning: Something went wrong: At least one minimal value of the aggregated time series exceeds the minimal value the input time series
>   warnings.warn("Something went wrong: At least one minimal value of the aggregated time series exceeds the minimal value the input time series")
> DatetimeIndex(['2011-01-01 00:00:00', '2011-01-01 01:00:00',
>                '2011-01-01 02:00:00', '2011-01-01 03:00:00',
>                '2011-01-01 04:00:00', '2011-01-01 05:00:00',
>                '2011-01-01 06:00:00', '2011-01-01 07:00:00',
>                '2011-01-01 08:00:00', '2011-01-01 09:00:00',
>                '2011-01-01 10:00:00', '2011-01-01 11:00:00',
>                '2011-01-01 12:00:00', '2011-01-01 13:00:00',
>                '2011-01-01 14:00:00', '2011-01-01 15:00:00',
>                '2011-01-01 16:00:00', '2011-01-01 17:00:00',
>                '2011-01-01 18:00:00', '2011-01-01 19:00:00',
>                '2011-01-01 20:00:00', '2011-01-01 21:00:00',
>                '2011-01-01 22:00:00', '2011-01-01 23:00:00',
>                '2011-01-03 00:00:00', '2011-01-03 01:00:00',
>                '2011-01-03 02:00:00', '2011-01-03 03:00:00',
>                '2011-01-03 04:00:00', '2011-01-03 05:00:00',
>                '2011-01-03 06:00:00', '2011-01-03 07:00:00',
>                '2011-01-03 08:00:00', '2011-01-03 09:00:00',
>                '2011-01-03 10:00:00', '2011-01-03 11:00:00',
>                '2011-01-03 12:00:00', '2011-01-03 13:00:00',
>                '2011-01-03 14:00:00', '2011-01-03 15:00:00',
>                '2011-01-03 16:00:00', '2011-01-03 17:00:00',
>                '2011-01-03 18:00:00', '2011-01-03 19:00:00',
>                '2011-01-03 20:00:00', '2011-01-03 21:00:00',
>                '2011-01-03 22:00:00', '2011-01-03 23:00:00'],
>               dtype='datetime64[ns]', name='snapshot', freq=None)
> INFO:pypsa.opf:Performed preliminary steps
> ```
> 
> I tried to cluster the first 72h to 2 days.

I looked into this one more time, although I had already checked this issue during working on my thesis. In tsam it is checked if the "new" timeseries exceed / fall below the maximum / minimum values of the original timeseries at any time. This happens in some cases, but only in no longer relevant decimal places (in your example 14th decimal place). The cause is probably the normalization and backscaling. I think we can ignore this, but the warning will pop up here and there....

File: etrago/appl.py [hunk 1]
```diff
@@ -121,17 +121,19 @@
     },
     "network_clustering_ehv": False,  # clustering of HV buses to EHV buses.
     "disaggregation": "uniform",  # None, 'mini' or 'uniform'
+    # Temporal Complexity:
     "snapshot_clustering": {
-        "active": False,  # choose if clustering is activated
+        "active": True,  # choose if clustering is activated
```
Comment (ClaraBuettner) [inline etrago/appl.py:126]: The default settings are now clustering the first two hours of the year to 5 days and skip every 3rd hour. Maybe we should change this a bit.

Comment (ClaraBuettner) [reply etrago/appl.py]: Thank you!

**Review by ClaraBuettner — APPROVED** [review]

**Pull request merged**